src/ray/util/process.h

```diff
@@ -0,0 +1,131 @@
+#ifndef RAY_UTIL_PROCESS_H
+#define RAY_UTIL_PROCESS_H
+
+#ifdef __linux__
+#include <fcntl.h>     // O_RDONLY
+#include <sys/stat.h>  // open()
+#include <unistd.h>    // close(), dup()
+#endif
+
+#include <algorithm>   // std::swap
+#include <functional>  // std::equal_to, std::hash, std::less
+#include <memory>      // std::shared_ptr
+#include <utility>     // std::std::forward
+
+#include <boost/asio/io_service.hpp>
+#include <boost/process/args.hpp>
+#include <boost/process/child.hpp>
+
+// We only define operators required by the standard library (<, ==, hash).
+// We declare but avoid defining the rest so that they're not used by accident.
+
+namespace ray {
+
+typedef boost::process::pid_t pid_t;
+
+class Process : public boost::process::child {
+  typedef boost::process::child base_type;
+
+ protected:
+  class ProcessFD {
+    // This class makes a best-effort attempt to keep a PID alive.
+    // However, it cannot make any guarantees.
+    // The kernel might not even support this mechanism.
+    // See here: https://unix.stackexchange.com/a/181249
+#ifdef __linux__
+    int fd_;
+#endif
+   public:
+#ifdef __linux__
+    ~ProcessFD() {
+      if (fd_ != -1) {
+        ::close(fd_);
+      }
+    }
+    ProcessFD(pid_t pid) : fd_(-1) {
+      if (pid != -1) {
+        char path[64];
+        sprintf(path, "/proc/%d/ns/pid", static_cast<int>(pid));
+        fd_ = ::open(path, O_RDONLY);
+      }
+    }
+    ProcessFD(ProcessFD &&other) : fd_(std::move(other.fd_)) { other.fd_ = -1; }
+    ProcessFD(const ProcessFD &other) : fd_(other.fd_ != -1 ? ::dup(other.fd_) : -1) {}
+    ProcessFD &operator=(ProcessFD other) {
+      using std::swap;
```

Comment:
remove this line and use std::swap directly

Reply:
I can do it (it makes no difference for `int`), but `std::swap` isn't the right way to call `swap`. The pattern for calling swap is `using std::swap; swap(...);` [to enable ADL](https://en.cppreference.com/w/cpp/language/adl#Notes).

Reply:
Feel free to keep it. I don't think it is super important here though (because these are ints) but it is good for things to not break in the future.